akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala

```diff
@@ -91,11 +106,72 @@ private[internal] object TcpDnsClient {
   def decodeLength(data: ByteString): Int =
     ((data(0).toInt + 256) % 256) * 256 + ((data(1) + 256) % 256)
 
-  def throwFailure(message: String, cause: Option[Throwable]): Unit =
+  def throwFailure(message: String, cause: Option[Throwable], log: LoggingAdapter, reportTo: ActorRef): Nothing = {
     cause match {
       case None =>
+        log.warning("TCP DNS client failed: {}", message)
+        reportTo ! DnsClient.TcpDropped
         throw new AkkaException(message)
+
       case Some(throwable) =>
+        log.warning(throwable, "TCP DNS client failed: {}", message)
+        reportTo ! DnsClient.TcpDropped
         throw new AkkaException(message, throwable)
     }
+  }
+
+  abstract class Writer(val connection: ActorRef, val log: LoggingAdapter) {
+    def maybeWriteMessage(msg: Message): Writer
+    def ack(): Writer
+
+    protected def writeMessage(msg: Message): Unit = {
+      val bytes = msg.write()
+      connection ! Tcp.Write(encodeLength(bytes.length) ++ bytes, Ack)
+    }
+  }
+
+  class Writing(connection: ActorRef, log: LoggingAdapter) extends Writer(connection, log) {
+    def maybeWriteMessage(msg: Message): Writer = {
+      writeMessage(msg)
+
+      new Buffering(connection, log)
```

Comment:
I'm not sure we really want to allow just the one in flight write message between this actor and the TCP manager

Reply:
Ok, that's actually how we show the ack based protocol in the docs, so might be fine.

Reply:
Yeah the TCP manager only allows one write at a time.  I originally went for the "NACK-based with write-suspending" approach (assume the write will be accepted until you get the backpressure signal, then resend the ones that were backpressured), but the complexity there got a bit out of hand and I wasn't sure that actually gave us that much.